src/connection/host_lookup.c

```diff
@@ -0,0 +1,528 @@
+ *
+ */
+
+/*
+ * host_lookup.c
+ */
+
+#ident "$Id$"
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <unordered_map>
+
+#if defined (WINDOWS)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <arpa/inet.h>
+#include <netdb.h>
+#endif
+
+#include "system_parameter.h"
+#include "environment_variable.h"
+
+#define HOSTNAME_BUF_SIZE              (256)
+#define MAX_HOSTS_LINE_NUM       (256)
+#define IPADDR_LEN              (17)
+
+typedef enum
+{
+  HOSTNAME_TO_IPADDR = 0,
+  IPADDR_TO_HOSTNAME = 1,
+} LOOKUP_TYPE;
+
+typedef enum
+{
+  INSERT_IPADDR = 0,
+  INSERT_HOSTNAME = 1,
+} HOSTENT_INSERT_TYPE;
+
+typedef enum
+{
+  USE_GLIBC_HOSTS = 0,
+  USE_USER_DEFINED_HOSTS = 1,
+} HOST_LOOKUP_TYPE;
+
+typedef enum
+{
+  HP = 0,
+  H_NAME,
+  H_ALIASES,
+  H_ADDR_LIST,
+  H_ALIASES0,
+  H_ADDR_LIST0,
+} HOSTENT_ELEMENT;
+
+struct cub_hostent
+{
+  char ipaddr[IPADDR_LEN];
+  char hostname[HOSTNAME_BUF_SIZE + 1];
+};
+
+typedef struct cub_hostent CUB_HOSTENT;
+static CUB_HOSTENT hostent_List[MAX_HOSTS_LINE_NUM];	/* "etc/hosts" file maximum line */
+static const char user_defined_hostfile_Name[] = "hosts.conf";
+
+static int host_conf_element_Count = 0;
+
+// *INDENT-OFF*
+static std::unordered_map <std::string, std::string> user_host_Map;
+// *INDENT-ON*
+
+static struct hostent *hostent_init ();
+static int host_conf_load ();
+static struct hostent *host_lookup_internal (const char *hostname, struct sockaddr *saddr, LOOKUP_TYPE lookup_case);
+
+static struct hostent *
+hostent_init ()
+{
+  struct hostent *hp;
+  HOSTENT_ELEMENT hostent_element;
+  int hostent_alloc_order;
+
+  if ((hp = (struct hostent *) malloc (sizeof (struct hostent))) == NULL)
+    {
+      hostent_alloc_order = HP;
+      goto FREE_;
+    }
+  if ((hp->h_name = (char *) malloc (sizeof (char) * (HOSTNAME_BUF_SIZE + 1))) == NULL)
+    {
+      hostent_alloc_order = H_NAME;
+      goto FREE_;
+    }
+  if ((hp->h_aliases = (char **) malloc (sizeof (char *))) == NULL)
+    {
+      hostent_alloc_order = H_ALIASES;
+      goto FREE_;
+    }
+  if ((hp->h_addr_list = (char **) malloc (sizeof (char *) * HOSTNAME_BUF_SIZE)) == NULL)
+    {
+      hostent_alloc_order = H_ADDR_LIST;
+      goto FREE_;
+    }
+  if ((hp->h_aliases[0] = (char *) malloc (sizeof (char) * (HOSTNAME_BUF_SIZE + 1))) == NULL)
+    {
+      hostent_alloc_order = H_ALIASES0;
+      goto FREE_;
+    }
+  if ((hp->h_addr_list[0] = (char *) malloc (sizeof (char) * IPADDR_LEN)) == NULL)
+    {
+      hostent_alloc_order = H_ADDR_LIST0;
+      goto FREE_;
+    }
+  hp->h_addrtype = AF_INET;
+  hp->h_length = 4;
+
+  return hp;
+
+FREE_:
+  switch (hostent_alloc_order)
+    {
+    case H_ADDR_LIST0:
+      free (hp->h_aliases[0]);
+
+    case H_ALIASES0:
+      free (hp->h_addr_list);
+
+    case H_ADDR_LIST:
+      free (hp->h_aliases);
+
+    case H_ALIASES:
+      free (hp->h_name);
+
+    case H_NAME:
+      free (hp);
+
+    }
+
```

Comment:
```suggestion
  if (hp)
  {
    FREE_MEM (hp->h_addr_list)
    FREE_MEM (hp->h_aliases);
    FREE_MEM (hp->h_name);
    FREE_MEM (hp);
  }
```

Reply:
We cannot use FREE_MEM macro, if the cci_common.h is included. Build cannot be done.

Because I guess that as much like db_error_string(), there are functions or macro that cannot use in the server environment.